### PR TITLE
[pkg/ottl] Add keep_keys slice argument coverage

### LINKED ISSUE #45980: Support slice argument in keep_keys
### Component(s)

pkg/ottl

### Is your feature request related to a problem? Please describe.

Currently, it seems like the `keep_keys` function in OTTL does not support a slice argument as the second parameter

For example

```yaml
- set(cache["list_of_keys"], ["test"])
- keep_keys(datapoint.attributes, cache["list_of_keys"])
```

Will throw 

```
unable to run metrics configuration. Error: unable to parse OTTL statement "keep_keys(datapoint.attributes, datapoint.cache[\"list_of_keys\"])": error while parsing arguments for call to "keep_keys": invalid argument at position 1: must be a list of type StringGetter[*github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/contexts/ottldatapoint.TransformContext]
```

### Describe the solution you'd like

To adapt `keep_keys` to be able to do `keep_keys(datapoint.attributes, cache["list_of_keys"])`, maybe using `ottl.PSliceGetter` ?

### Describe alternatives you've considered

I was not able to find any working bypass

### Additional context

#38690 is a similar issue with the `Concat()` converter

### Tip

<sub>[React](https://github.blog/news-insights/product-news/add-reactions-to-pull-requests-issues-and-comments/) with 👍 to help prioritize this issue. Please use comments to provide useful context, avoiding `+1` or `me too`, to help us triage it. Learn more [here](https://opentelemetry.io/community/end-user/issue-participation/).</sub>

### PR DESCRIPTION
**Pull request by RutamBhagat**
#### Description

This PR updates keep_keys to accept its list of keys from list literals, paths, and converter expressions

keep_keys previously declared Keys as []ottl.StringGetter[K] this type works for literal lists like ["foo", "bar"], but the OTTL parser cant use a path or converter expression that returns the entire slice. For example, keep_keys(log.attributes, cache["keys"]) will fail when parsing the configuration, before keep_keys can be executed

##### Why does it fail?

For a bare Go slice argument, OTTL expects a literal-list AST node. It processes each list element during configuration parsing and wraps it in a StringGetter. A path or converter is a dynamic expression not a literal-list node. Thats why, it cant use this argument-building path, even when it returns a valid string slice

##### How does ottl.SliceGetter fix it?

KeepKeysArguments.Keys now uses ottl.SliceGetter[K, ottl.StringGetter[K]] the Literal lists continue to use the existing parse-time behavior. SliceGetter resolves dynamic slices at runtime and validates each element through StringGetter. Then, keep_keys builds the key set and filters the target map

The current literal-key optimization is preserved. When SliceGetter reports a known length and all values are literals, keep_keys precomputes the key set once. The length check stops literal extraction from evaluating a runtime path with a zero-value transform context

This changes the exported KeepKeysArguments.Keys field type. Downstream Go callers that construct KeepKeysArguments must update this field. Existing valid OTTL statements that pass literal lists stay unchanged

#### Link to tracking issue

Fixes #45980

#### Testing

Ran `cd pkg/ottl && go test . ./ottlfuncs ./e2e -count=1`

Updated the existing keep_keys unit tests for the new argument type, added an end-to-end regression test for a cached key list, and regenerated the function-signature golden file

#### Documentation

Updated pkg/ottl/ottlfuncs/README.md to describe dynamic slice arguments. Added examples that use Split and a cached key list

#### Authorship

- [x] I, a human, wrote this pull request description myself.

### CODE REVIEW

**Comment by RutamBhagat** [comment]
not sure why the 1 ci failed, doesnt seems to be related

File: pkg/ottl/ottlfuncs/func_keep_keys.go [hunk 1]
```diff
@@ -49,19 +49,25 @@ func keepKeys[K any](target ottl.PMapGetSetter[K], keys []ottl.StringGetter[K])
...
-			// Build key set at runtime for dynamic keys
-			keySet = make(map[string]struct{}, len(keys))
-			for _, key := range keys {
-				k, err := key.Get(ctx, tCtx)
-				if err != nil {
-					return nil, err
+		keySet := literalKeySet
+		if keySet == nil {
+			// Resolve dynamic or runtime-generated keys for the current transform context
+			keySet = make(map[string]struct{})
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_keep_keys.go]: We need to keep the `len(keys)` value in this make call that we had previously. It makes this function a lot more performant.

File: pkg/ottl/ottlfuncs/func_keep_keys_test.go
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_keep_keys_test.go:1]: Can you make sure all these test cases are included:

- `keep_keys(attributes, cache["x"])` where `cache["x"]` is unset/nil
- `keep_keys(attributes, cache["x"])` where `cache["x"]` is a non-slice (e.g. string/int)
- Dynamic slice containing a non-string element
- Dynamic key getter that returns an error
- Empty dynamic slice (keep none)
- Literal-keys path asserted via a parsed `["a","b"]` statement

File: pkg/ottl/ottlfuncs/func_keep_keys.go (hunk 1 above)
Comment (RutamBhagat) [reply pkg/ottl/ottlfuncs/func_keep_keys.go]: fixed the perf and memory regression

File: pkg/ottl/ottlfuncs/func_keep_keys_test.go
Comment (RutamBhagat) [reply pkg/ottl/ottlfuncs/func_keep_keys_test.go:1]: added more unit and e2e tests

File: pkg/ottl/e2e/e2e_test.go [hunk 2]
```diff
@@ -2323,6 +2323,90 @@ func Test_e2e_ottl_statement_sequence(t *testing.T) {
...
+	}{
+		{
+			name:      "valid dynamic slice",
+			setup:     `set(cache["x"], ["flags"])`,
+			statement: `keep_keys(attributes, cache["x"])`,
+			wantKeys:  []string{"flags"},
+		},
+		{
+			name:      "unset dynamic slice keeps none",
+			statement: `keep_keys(attributes, cache["x"])`,
```
Comment (TylerHelmuth) [inline pkg/ottl/e2e/e2e_test.go]: Is the behavior of `cache["x"]` retuning till to drop everything from `attributes`? That doesn't feel right to me. @edmocosta @evan-bradley what do you think?

File: .chloggen/ottl-keep-keys-slice-getter-api.yaml
Comment (TylerHelmuth) [inline .chloggen/ottl-keep-keys-slice-getter-api.yaml:1]: now that the function args are unexporter this changelog can be deleted.

File: pkg/ottl/e2e/e2e_test.go (hunk 2 above)
Comment (TylerHelmuth) [reply pkg/ottl/e2e/e2e_test.go]: This is the behavior keep_keys already had if you passed in `[]`. I feel like `nil` should behave different. A `[]` is an explicit request to keep no keys. I think the functino should error if it gets `nil` for the keys list.

Comment (RutamBhagat) [reply pkg/ottl/e2e/e2e_test.go]: should `SliceGetter.Range` return a type error if the runtime expression resolves to untyped `nil`? and should `SliceGetter.Get` use the same behavior for consistency? explicit empty slices would remain valid

Comment (RutamBhagat) [reply pkg/ottl/e2e/e2e_test.go]: #51174  created an issue for this here

Comment (RutamBhagat) [reply pkg/ottl/e2e/e2e_test.go]: updated `keep_keys` to use `SliceGetter.Get` instead of `Range`

`Get` keeps the difference between `nil` and an empty slice

unset cache["x"] now returns an error and [] keeps existing behavior and removes all attributes

I have also closed #51174 because this PR does not require a shared SliceGetter change

File: .chloggen/ottl-keep-keys-slice-getter-api.yaml
Comment (RutamBhagat) [reply .chloggen/ottl-keep-keys-slice-getter-api.yaml:1]: removed this file

File: pkg/ottl/e2e/e2e_test.go [hunk 3]
```diff
@@ -2323,6 +2323,91 @@ func Test_e2e_ottl_statement_sequence(t *testing.T) {
 	}
 }
 
+func Test_e2e_keep_keys_slice_arguments(t *testing.T) {
```
Comment (TylerHelmuth) [inline pkg/ottl/e2e/e2e_test.go]: All these test cases should be covered in the function-specfic test. For the e2e test it is sufficient to add a new entry to the existing list that passes the cache for the list.

File: pkg/ottl/ottlfuncs/func_keep_keys.go [hunk 4]
```diff
@@ -51,14 +52,19 @@ func keepKeys[K any](target ottl.PMapGetSetter[K], keys []ottl.StringGetter[K])
...
-			// Use pre-built key set for literal keys
-			keySet = literalKeySet
-		} else {
-			// Build key set at runtime for dynamic keys
-			keySet = make(map[string]struct{}, len(keys))
-			for _, key := range keys {
+		keySet := literalKeySet
+		if keySet == nil {
+			// Resolve dynamic or runtime-generated keys for the current transform context
+			resolvedKeys, err := keys.Get(ctx, tCtx)
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_keep_keys.go:58]: We have an issue where a statement like `keep_keys(attributes, nil)` causes a panic. this is a pretty silly scenario, but panics are bad. 

I believe this is a bug in the SliceGetter API. I think `Get` needs to have a nil check before it uses `runtimeSlice`

File: pkg/ottl/e2e/e2e_test.go (hunk 3 above)
Comment (RutamBhagat) [reply pkg/ottl/e2e/e2e_test.go]: moved the tests to `pkg/ottl/ottlfuncs/func_keep_keys_test.go` except the new e2e cache slice test

**Comment by TylerHelmuth** [comment]
@RutamBhagat I merged in a fix for the nil issue in SliceGetter

**Comment by RutamBhagat** [comment]
I pulled in the fix and added a `keep_keys(attributes, nil)` regression test to cover this edge case

File: pkg/ottl/ottlfuncs/func_keep_keys.go (hunk 4 above)
Comment (RutamBhagat) [reply pkg/ottl/ottlfuncs/func_keep_keys.go:58]: added new test to cover this case

File: pkg/ottl/e2e/e2e_test.go (hunk 2 above)
Comment (edmocosta) [reply pkg/ottl/e2e/e2e_test.go]: > Is the behavior of cache["x"] retuning till to drop everything from attributes? That doesn't feel right to me. @edmocosta @evan-bradley what do you think?

IMO,  if `cache["x"]` is `nil`, raise an error as @TylerHelmuth suggested. Making `nil` fallback to `[]`  will make this function as confusing as the `aggregate_on_attributes`, which we already discussed in the past: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/45789

File: pkg/ottl/ottlfuncs/func_keep_keys.go [hunk 5]
```diff
@@ -30,19 +30,20 @@ func createKeepKeysFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments)
...
-	for _, key := range keys {
-		k, isLiteral := ottl.GetLiteralValue(key)
-		if !isLiteral {
-			literalKeySet = nil
-			break
+func keepKeys[K any](target ottl.PMapGetSetter[K], keys *ottl.SliceGetter[K, ottl.StringGetter[K]]) ottl.ExprFunc[K] {
+	// Pre-build the key set when the slice length and all values are known at parse time
+	var literalKeySet map[string]struct{}
+
+	if length, known := keys.Len(); known {
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_keep_keys.go]: We shouldn't need to call `Len` here. I've opened a [PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/51438) fixing that so we can drop it, and only use the `ottl.GetLiteralValues`.

File: pkg/ottl/e2e/e2e_test.go (hunk 2 above)
Comment (RutamBhagat) [reply pkg/ottl/e2e/e2e_test.go]: current implementation has the requested behavior in unset cache entry test

```go
{
    name:        "unset cache entry",
    statement:   `keep_keys(attributes, cache["x"])`,
    wantErrPart: "keys cannot be nil",
}
```

File: pkg/ottl/ottlfuncs/func_keep_keys.go (hunk 5 above)
Comment (RutamBhagat) [reply pkg/ottl/ottlfuncs/func_keep_keys.go]: will update once merged

Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_keep_keys.go]: Merged

Comment (RutamBhagat) [reply pkg/ottl/ottlfuncs/func_keep_keys.go]: dropped the Len() check and now only uses ottl.GetLiteralValues

**Review by TylerHelmuth — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
@RutamBhagat no need to merge in the latest main each time. It means rerunning all the CI and that takes forever and will make it harder to merge this PR.

**Comment by RutamBhagat** [comment]
> @RutamBhagat no need to merge in the latest main each time. It means rerunning all the CI and that takes forever and will make it harder to merge this PR.

ok sorry didnt know that

File: pkg/ottl/ottlfuncs/func_keep_keys.go [hunk 6]
```diff
@@ -30,19 +30,18 @@ func createKeepKeysFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments)
...
-	for _, key := range keys {
-		k, isLiteral := ottl.GetLiteralValue(key)
-		if !isLiteral {
-			literalKeySet = nil
-			break
+func keepKeys[K any](target ottl.PMapGetSetter[K], keys *ottl.SliceGetter[K, ottl.StringGetter[K]]) ottl.ExprFunc[K] {
+	// Pre-build literal keys, but let nil reach the runtime check, this preserves the error for nil without treating it as an empty slice
+	var literalKeySet map[string]struct{}
+
+	if literalValues, allLiteral := ottl.GetLiteralValues[K, string](keys); allLiteral && literalValues != nil {
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_keep_keys.go]: If `allLiteral` is `true`, and `literalValues` is `nil`, we could return an error here and not postpone the failure to the function's execution.

File: pkg/ottl/ottlfuncs/func_keep_keys.go [hunk 7]
```diff
@@ -51,14 +50,19 @@ func keepKeys[K any](target ottl.PMapGetSetter[K], keys []ottl.StringGetter[K])
...
-			// Use pre-built key set for literal keys
-			keySet = literalKeySet
-		} else {
-			// Build key set at runtime for dynamic keys
-			keySet = make(map[string]struct{}, len(keys))
-			for _, key := range keys {
+		keySet := literalKeySet
+		if keySet == nil {
+			// Resolve dynamic or runtime-generated keys for the current transform context
+			resolvedKeys, err := keys.Get(ctx, tCtx)
```
Comment (edmocosta) [inline pkg/ottl/ottlfuncs/func_keep_keys.go]: Could you please change it to use `.Range` instead?  `.Get` internally loops over the slice to collect all values, and this function loop over it again at line 65. 

The `nil` use-case isn't supported by `.Range`, but I think we could double check that later on the process by calling `.Get()` if `len(keySet) == 0`.

Something like:

```go
			if size, ok := keys.Len(); ok {
				keySet = make(map[string]struct{}, size)
			}
			err := keys.Range(ctx, tCtx, func(key ottl.StringGetter[K]) bool {
				k, err := key.Get(ctx, tCtx)
				if err != nil {
					return false
				}
				keySet[k] = struct{}{}
				return true
			})
			if err != nil {
				return nil, err
			}
			if len(keySet) == 0 {
				nilVal, err := keys.Get(ctx, tCtx)
				if err != nil {
					return nil, err
				}
				if nilVal == nil {
					return nil, errors.New("keys cannot be nil")
				}
			}
```

WDYT? @TylerHelmuth ?

Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_keep_keys.go]: Ya this is worth not always having to loop twice for the most common cases.

I do think this is an indicator that we need to tweak the SliceGetter API somehow to handle this better.
